lib/Dialect/Handshake/Transforms/LowerExtmemToHW.cpp

```diff
@@ -0,0 +1,255 @@
+  void runOnOperation() override {
+    auto op = getOperation();
+    for (auto func : op.getOps<handshake::FuncOp>()) {
+      if (failed(LowerExtmemToHW(func))) {
+        signalPassFailure();
+        return;
+      }
+    }
+  };
+
+  LogicalResult LowerExtmemToHW(handshake::FuncOp func);
+};
+
+static Value plumbLoadPort(Location loc, OpBuilder &b,
+                           const handshake::ExtMemLoadInterface &ldif,
+                           Value loadData) {
+  // We need to feed both the load data and the load done outputs.
+  // Fork the extracted load data into two, and 'join' the second one to
+  // generate a none-typed output to drive the load done.
+  auto dataFork = b.create<ForkOp>(loc, loadData, 2);
+
+  auto dataOut = dataFork.getResult()[0];
+  llvm::SmallVector<Value> joinArgs = {dataFork.getResult()[1]};
+  auto dataDone = b.create<JoinOp>(loc, joinArgs);
+
+  ldif.dataOut.replaceAllUsesWith(dataOut);
+  ldif.doneOut.replaceAllUsesWith(dataDone);
+
+  // Return load address, to be fed to the top-level output.
+  return ldif.addressIn;
+}
+
+static Value plumbStorePort(Location loc, OpBuilder &b,
+                            const handshake::ExtMemStoreInterface &stif,
+                            Value done, Type outType) {
+  stif.doneOut.replaceAllUsesWith(done);
+  // Return the store and data to be fed to the top-level output.
+  llvm::SmallVector<Value> structArgs = {stif.dataIn, stif.addressIn};
+  return b
+      .create<hw::StructCreateOp>(loc, outType.cast<hw::StructType>(),
+                                  structArgs)
+      .getResult();
+}
+
+static void appendToStringArrayAttr(Operation *op, StringRef attrName,
+                                    StringRef attrVal) {
+  auto *ctx = op->getContext();
+  llvm::SmallVector<Attribute> newArr;
+  llvm::copy(op->getAttrOfType<ArrayAttr>(attrName).getValue(),
+             std::back_inserter(newArr));
+  newArr.push_back(StringAttr::get(ctx, attrVal));
+  op->setAttr(attrName, ArrayAttr::get(ctx, newArr));
+}
+
+static void insertInStringArrayAttr(Operation *op, StringRef attrName,
+                                    StringRef attrVal, unsigned idx) {
+  auto *ctx = op->getContext();
+  llvm::SmallVector<Attribute> newArr;
+  llvm::copy(op->getAttrOfType<ArrayAttr>(attrName).getValue(),
+             std::back_inserter(newArr));
+  newArr.insert(newArr.begin() + idx, StringAttr::get(ctx, attrVal));
+  op->setAttr(attrName, ArrayAttr::get(ctx, newArr));
+}
+
+static void eraseFromArrayAttr(Operation *op, StringRef attrName,
+                               unsigned idx) {
+  auto *ctx = op->getContext();
+  llvm::SmallVector<Attribute> newArr;
+  llvm::copy(op->getAttrOfType<ArrayAttr>(attrName).getValue(),
+             std::back_inserter(newArr));
+  newArr.erase(newArr.begin() + idx);
+  op->setAttr(attrName, ArrayAttr::get(ctx, newArr));
+}
+
+LogicalResult
+HandshakeLowerExtmemToHWPass::LowerExtmemToHW(handshake::FuncOp func) {
```

Comment:
I usually defined a function with handles a region, instead of a `handshake::FuncOp`, as this helped me with my streaming work. On the other hand, if someone really requires this, its a brief change to implement.

Reply:
I think it'd be fair to assume a `handshake::FuncOp` here due to also needing to update in/out names.

> On the other hand, if someone really requires this, its a brief change to implement

+1.